File: android/app/src/main/java/app/inku/mobile/data/DdlExport.kt
package app.inku.mobile.data

import app.inku.mobile.pipeline.ImportedPluginDefinition
import app.inku.mobile.pipeline.pluginVisibleNames
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * Portable DDL: the visible DDL with the plugin definitions it names
 * (`inku.ddl-export.v1`). Mirrors the server's `ddl_export.build_ddl_export`
 * and the web's `parseDdlImport`. Definitions are data-only and are validated
 * again by the shared Rust boundary when a new work resolves them.
 */
object DdlExport {
    const val SCHEMA = "inku.ddl-export.v1"
    private const val MAX_IMPORTED_PLUGINS = 64

    /** Keeps only the definitions whose canonical name or an alias the DDL writes. */
    fun build(
        source: String,
        language: String,
        definitions: JSONArray?,
        summaries: JSONArray?,
        exportedFrom: JSONObject,
    ): JSONObject {
        val plugins = JSONArray()
        val seen = mutableSetOf<String>()
        for (index in 0 until (definitions?.length() ?: 0)) {
            val definition = definitions?.optJSONObject(index) ?: continue
            val names = pluginVisibleNames(JSONArray().put(definition))
            val name = names.firstOrNull() ?: continue
            if (name in seen || names.none { source.contains(it) }) continue
            seen += name
            plugins.put(
                JSONObject()
                    .put("definition", definition)
                    .put("summary", summaries?.opt(index) as? String ?: ""),
            )
        }
        return JSONObject()
            .put("schema", SCHEMA)
            .put("language", language)
            .put("ddl", source)
            .put("plugins", plugins)
            .put("exported_from", exportedFrom)
    }

    data class Import(
        val ddl: String,
        val plugins: List<ImportedPluginDefinition>,
        /** Canonical names of the carried plugins, for the author-facing notice. */
        val names: List<String>,
    )

    /** Any text that is not an export -- JSON or not -- is the author's DDL as written. */
    fun parse(text: String): Import {
        val file = try {
            JSONObject(text)
        } catch (_: JSONException) {
            return Import(text, emptyList(), emptyList())
        }
        if (file.optString("schema") != SCHEMA) return Import(text, emptyList(), emptyList())
        val ddl = file.opt("ddl") as? String ?: throw IllegalArgumentException("ddl_export_without_ddl")
        val items = file.optJSONArray("plugins") ?: JSONArray()
        require(items.length() <= MAX_IMPORTED_PLUGINS) { "ddl_export_too_many_plugins" }
        val plugins = (0 until items.length()).map { index ->
            val definition = items.optJSONObject(index)?.optJSONObject("definition")
                ?: throw IllegalArgumentException("ddl_export_invalid_plugin")
            val summary = (items.optJSONObject(index)?.opt("summary") as? String)?.trim().orEmpty()
            ImportedPluginDefinition(
                definitionJson = definition.toString(),
                // Stage 1 requires a summary; an export without one still names the plugin.
                summary = summary.ifEmpty { "${definition.optString("namespace")}.${definition.optString("heading")}" },
            )
        }
        val names = plugins.map { plugin ->
            JSONObject(plugin.definitionJson).let { "${it.optString("namespace")}.${it.optString("heading")}" }
        }
        return Import(ddl, plugins, names)
    }
}

File: server/src/inku_server/ddl_export.py
"""Portable DDL export: the visible DDL with the plugin definitions it uses.

A saved work already keeps its exact plugin definitions, so it redraws the same
way on this server. Plain DDL text does not: another server may lack a plugin or
hold a different edition. An export carries the definitions the DDL names so the
reading side can use them for that work alone (see macro_catalog imported
definitions). Definitions are data-only `inku.macro-definition.v1` objects and
are validated again by the shared Rust boundary when they are read.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

DDL_EXPORT_SCHEMA = "inku.ddl-export.v1"


def _qualified_name(definition: Mapping[str, Any]) -> str | None:
    namespace, heading = definition.get("namespace"), definition.get("heading")
    return f"{namespace}.{heading}" if isinstance(namespace, str) and isinstance(heading, str) else None


def _visible_names(definition: Mapping[str, Any]) -> list[str]:
    """The canonical qualified name and its aliases; the DDL may write any of them."""
    name = _qualified_name(definition)
    if name is None:
        return []
    aliases = definition.get("aliases") or []
    return [name, *(f"{definition['namespace']}.{alias}" for alias in aliases if isinstance(alias, str))]


def build_ddl_export(
    source: str,
    language: str,
    definitions: Sequence[Mapping[str, Any]] = (),
    summaries: Sequence[str] = (),
    exported_from: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Keep only the definitions whose qualified name the DDL writes."""
    plugins = []
    seen: set[str] = set()
    for index, definition in enumerate(definitions):
        name = _qualified_name(definition)
        if name is None or name in seen or not any(visible in source for visible in _visible_names(definition)):
            continue
        seen.add(name)
        summary = summaries[index] if index < len(summaries) and isinstance(summaries[index], str) else ""
        plugins.append({"definition": dict(definition), "summary": summary})
    return {
        "schema": DDL_EXPORT_SCHEMA,
        "language": language,
        "ddl": source,
        "plugins": plugins,
        "exported_from": dict(exported_from or {}),
    }
